src/Microsoft.AspNetCore.Routing/Matchers/AsciiKeyedJumpTable.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    // An optimized jump table that trades a small amount of additional memory for
+    // hash-table like performance.
+    //
+    // The optimization here is to use the first character of the known entries
+    // as a 'key' in the hash table in the space of A-Z. This gives us a maximum
+    // of 26 buckets (hence the reduced memory)
+    internal class AsciiKeyedJumpTable : JumpTable
```

Comment:
BTW - I'm not totally confident in this guy yet, I'm not sure that it's completely worth doing vs other approaches. Basically, I think it needs a little more investigation...

Reply:
It's better than a linear search, once you have 3-4 candidates, but worse than approaches based on true hash tables.

Reply:
Could make it adaptive? Different strategies depending on the number of candidates

Reply:
Yes, it is actually, see `JumpTableMatcherBuilder`. What I mean is that is that the sweet spot for this is really small - it may not be worth the complexity.

Additionally, this looks really good on small artificial benchmarks, but offers much less benefit for something like Azure or github's route table. 

If we ended up with an adaptive builder where:
0: special case
1: special case
2-4: linear search
5-8??: ASCII
9+: Dictionary

Then it seems worth doing. I suspect if we put more time into it, the dictionary/hash based implementation will have it's floor pushed down way further (potentially to 2-3).